godot-core/src/builtin/color_hsv.rs

```diff
@@ -0,0 +1,278 @@
+/// 4 times faster - partly because it avoids calls between Godot and Rust during conversion.
+///
+/// ## Conversions
+///
+/// Both conversions (`Color` to `ColorHsv` and `ColorHsv` to `Color`) will panic if RGBA or HSVA values are not within range `0.0..=1.0`.
+/// To ensure the values are in valid range, methods [Color::normalize()] and [ColorHsv::normalize_clamped_h()] or [ColorHsv::normalize_wrapped_h()]
+/// can be used.
+///
+/// ```
+/// use godot::builtin::{Color, ColorHsv};
+///
+/// let rgb = Color { r: 1.15, ..Default::default() };
+/// let hsv = ColorHsv { h: 1.15, ..Default::default() };
+///
+/// // Below conversions would panic.
+/// // let hsv_from_rgb: ColorHsv = rgb.into();
+/// // let rgb_from_hsv: Color = hsv.into();
+///
+/// let hsv_from_rgb: ColorHsv = rgb.normalize().into();
+/// let rgb_from_hsv: Color = hsv.normalize_wrapped_h().into();
+/// ```
+///
+/// ## Precision warning
+/// Conversions between `f32`-based RGB and HSV representations are not completely lossless. Try to avoid repeatable
+/// `Color` -> `ColorHsv` -> `Color` roundtrips. One way to minimalize possible distortions is to keep `ColorHsv` on the Rust side, apply
+/// all changes to this struct and convert it to `Color` before moving to the Godot side, instead of fetching `Color` from Godot side before
+/// every mutation, though changes should be minimal if color values are mutated either only on `Color` or `ColorHsv` side.
+///
+/// ## Examples
+///
+/// ```
+/// use godot::builtin::{Color, ColorHsv};
+///
+/// // ColorHsv can be constructed from only Hue, Saturation and Value.  
+/// let mut c_hsv = ColorHsv::from_hsv(0.74, 0.69, 0.18);
+///
+/// // Or with Alpha value also specified. If not specified, it is set at 1.0.
+/// let mut c_hsv2 = ColorHsv::from_hsva(0.74, 0.69, 0.18, 1.0);
+///
+/// assert_eq!(c_hsv, c_hsv2);
+///
+/// // Two way convertion: Color -> ColorHsv -> Color is not entirely lossless. Such repeatable
+/// // convertions should be avoided, as they could build up to significant values if values
+/// // are mutated both on `Color` and `ColorHsv`.
+/// let color1 = Color::from_rgb(0.74, 0.69, 0.18);
+/// let color2: Color = ColorHsv::from(color1).into();
+///
+/// let diff_r = (color1.r - color2.r).abs();
+/// let diff_g = (color1.g - color2.g).abs();
+/// let diff_b = (color1.b - color2.b).abs();
+///
+/// assert_ne!(color1, color2);
+///
+/// assert!(diff_r < 0.001);
+/// assert!(diff_g < 0.001);
+/// assert!(diff_b < 0.001);
+/// ```
+///  
+/// ## Reference
+/// - Smith, Alvy Ray. "Color gamut transform pairs." ACM Siggraph Computer Graphics 12.3 (1978): 12-19.
+#[derive(Copy, Clone, Debug)]
+pub struct ColorHsv {
+    pub h: f32,
+    pub s: f32,
+    pub v: f32,
+    pub a: f32,
+}
+
+// Manual implementation because Hue minimal value (0.) is equal to maximal value (1.)
+impl PartialEq for ColorHsv {
+    fn eq(&self, other: &Self) -> bool {
+        (self.h == other.h || (self.h - other.h).abs() == 1.)
```

Comment:
This assumes that the hue is normalized, right? It's not necessarily be the case with `h` being public. The same for `ApproxEq`.

Reply:
Good catch! I think I will remove the custom `PartialEq` implementation and will wrap the Hue on `ApproxEq` with suitable documentation

Reply:
Please no implicit wraps -- make an `assert`. It is better that users are aware of potential logic errors.